src/JustSaying/Messaging/Channels/IDownloadBuffer.cs

```diff
@@ -0,0 +1,109 @@
+
+        public DownloadBuffer(
+            int bufferLength,
+            ISqsQueue sqsQueue,
+            ILoggerFactory logger)
+        {
+            _bufferLength = bufferLength;
+            _sqsQueue = sqsQueue;
+            _logger = logger.CreateLogger<IDownloadBuffer>();
+            _channel = Channel.CreateBounded<IQueueMessageContext>(bufferLength);
+        }
+
+        public async Task Start(CancellationToken stoppingToken)
+        {
+            ChannelWriter<IQueueMessageContext> writer = _channel.Writer;
+            try
+            {
+                while (!stoppingToken.IsCancellationRequested)
+                {
+                    var canWrite = await WaitToWriteAsync(writer, stoppingToken).ConfigureAwait(false);
+                    if (!canWrite) break;
+
+                    IList<Message> messages;
+                    using (_logger.TimedOperation("Receiving messages from SQS for queue {QueueName}", _sqsQueue.QueueName))
+                    {
+                        messages = await _sqsQueue
+                           .GetMessages(_bufferLength, _requestMessageAttributeNames, stoppingToken)
+                           .ConfigureAwait(false);
+
+                        if (messages == null) continue;
+                    }
+
+                    foreach (var message in messages)
+                    {
+                        IQueueMessageContext messageContext = _sqsQueue.CreateQueueMessageContext(message);
+                        await writer.WriteAsync(messageContext).ConfigureAwait(false);
+                    }
+                }
+
+                _logger.LogInformation("Downloader for queue {QueueName} has completed, shutting down channel...", _sqsQueue.Uri);
+            }
+            finally
+            {
+                writer.Complete();
+            }
+        }
+
+        private async Task<bool> WaitToWriteAsync(ChannelWriter<IQueueMessageContext> writer, CancellationToken stoppingToken)
+        {
+            while (!stoppingToken.IsCancellationRequested)
+            {
+                CancellationTokenSource linkedCts = null;
+                try
+                {
+                    // we don't want to pass the stoppingToken here because
+                    // we want to process any messages queued messages before stopping
+                    using var timeoutToken = new CancellationTokenSource(TimeSpan.FromSeconds(2));
+                    linkedCts = CancellationTokenSource.CreateLinkedTokenSource(timeoutToken.Token, stoppingToken);
+
+                    using (_logger.TimedOperation("Downloader waiting for buffer to empty before writing"))
+                    {
+                        bool writePermitted = await writer.WaitToWriteAsync(linkedCts.Token);
+                        return writePermitted;
+                    }
+                }
+                catch (OperationCanceledException)
+                {
+                    // no space in channel, check again
```

Comment:
[This](https://github.com/gkinsman/JustSaying/blob/0c5b2e1889f8f97d82d588c3f25c5b21b98e8596/src/JustSaying/Messaging/Channels/IDownloadBuffer.cs#L91) line throws that exception if the linked token is cancelled. So this assumes it was the timeout not the stopping token, but will check the stopping token again at the start of the loop. 
I added the timeout in because there was a case where it was stuck waiting there, but it might not be needed any more. Will test and check if it's still needed.